src/Admin/AdminHelper.php

```diff
@@ -39,13 +40,57 @@ class AdminHelper
     private const FORM_FIELD_DELETE = '_delete';
 
     /**
-     * @var Pool
+     * NEXT_MAJOR: Remove this property.
+     *
+     * @var Pool|null
      */
     protected $pool;
 
-    public function __construct(Pool $pool)
+    /**
+     * @var PropertyAccessorInterface
+     */
+    private $propertyAccessor;
+
+    /**
+     * NEXT_MAJOR: Change signature for (PropertyAccessorInterface $propertyAccessor).
+     */
+    public function __construct($poolOrPropertyAccessor)
     {
-        $this->pool = $pool;
+        // NEXT_MAJOR: Remove this block.
+        if (!$poolOrPropertyAccessor instanceof Pool && !$poolOrPropertyAccessor instanceof PropertyAccessorInterface) {
+            throw new \TypeError(sprintf(
+                'Argument 1 passed to "%s()" must be either an instance of %s or %s, %s given.',
+                __METHOD__,
+                Pool::class,
+                PropertyAccessorInterface::class,
+                \is_object($poolOrPropertyAccessor) ? \get_class($poolOrPropertyAccessor) : \gettype($poolOrPropertyAccessor)
+            ));
+        }
+
+        // NEXT_MAJOR: Remove this block.
+        if ($poolOrPropertyAccessor instanceof Pool) {
+            @trigger_error(sprintf(
+                'Passing an instance of "%s" as the argument 1 for "%s()" is deprecated since'
+                .' sonata-project/admin-bundle 3.x and will throw a \TypeError in version 4.0.'
```

Comment:
```suggestion
                'Passing an instance of "%s" as argument 1 for "%s()" is deprecated since'
                .' sonata-project/admin-bundle 3.x and will throw a \TypeError error in version 4.0.'
```